packages/@lwc/integration-karma/test/component/LightningElement.render/index.spec.js

```diff
@@ -82,3 +83,10 @@ it('supports returning different templates', () => {
         expect(elm.shadowRoot.textContent).toBe('Template 2');
     });
 });
+
+it('throws an error when render() returns an invalid value', () => {
+    const elm = createElement('x-render-invalid', { is: RenderInvalid });
+    expect(() => document.body.appendChild(elm)).toThrowCallbackReactionErrorDev(
```

Comment:
```suggestion
    expect(() => document.body.appendChild(elm)).toThrowCallbackReactionError(
```

Sorry, I missed that there is also a non-dev-only error thrown below: 

https://github.com/salesforce/lwc/blob/36498bd0c51d5869ced7b933a50638d868d47c1b/packages/%40lwc/engine-core/src/framework/template.ts#L249-L256

It really feels like we should just put both of these in the same place and both with a similar error message. It should be safe to combine them, since you'll get an error either way if the `html` is not a registered template function.

Reply:
The `isTemplateRegistered` check is inside `runWithBoundaryProtection`, but the `isFunction` check is not. Does that change the nature of the error at all? I poked into the git history, but both checks seem to have been around for a long time in some capacity, so presumably they serve different purposes. (Or at least they did, at some point in time?)

Reply:
@wjhsf That's a good question. The one inside of the `runWithBoundaryProtection` functions differently, for two reasons:

1. It runs in both prod and dev mode
2. It can be caught in an `errorCallback` (due to the boundary protection)

For the most part, we consider prod mode to be "the truth." We've actually had several efforts in the past to remove dev-only errors, since a lot of customers only test in prod mode and don't care if dev mode is broken. (Dev-only errors should probably be reserved for truly catastrophic errors that are only expected to be caught in our own Karma tests.)

For that reason, I would suggest getting rid of the dev-only error entirely, and just improving the error message for the error inside `runWithBoundaryProtection`. This would have a few advantages:

1. No changes to historical prod behavior – if it's not a registered function, it's going to throw inside of `runWithBoundaryProtection`
2. Same behavior in prod and dev mode. So you can use `toThrowCallbackReactionError` instead of `toThrowCallbackReactionErrorDev`.